File: bot/output.py
"""Shared whitelist output file generator.

Used by both the Discord bot worker and the standalone web service
to generate Squad RemoteAdminList format output files.
"""
from __future__ import annotations

from typing import TYPE_CHECKING

from bot.config import log, WEB_FILE_SECRET, WEB_INTERNAL_URL

if TYPE_CHECKING:
    from bot.database import Database


async def generate_output_files(db: "Database", guild_id: int) -> dict[str, str]:
    """Generate all whitelist output files for a guild.

    Returns a dict of {filename: content} for each output file.
    The files are in Squad RemoteAdminList format:

        Group=Whitelist:reserve
        Group=Admin:kick,ban,chat,cameraman,immune,reserve

        Admin=76561198012345678:Whitelist // PlayerName
        Admin=76561198087654321:Admin // AdminPlayer

    Args:
        db: Database instance
        guild_id: The guild to generate files for

    Returns:
        Dict mapping filename -> file content string
    """
    from bot.utils import to_bool

    # Get output mode setting
    output_mode = await db.get_setting(guild_id, "output_mode", "combined")
    combined_filename = await db.get_setting(guild_id, "combined_filename", "whitelist.txt")
    dedupe = to_bool(await db.get_setting(guild_id, "duplicate_output_dedupe", "true"))

    # Get all whitelists for this guild
    whitelists = await db.get_whitelists(guild_id)
    whitelist_by_id = {wl["id"]: wl for wl in whitelists}

    # Get all squad groups for this guild
    squad_groups = await db.get_squad_groups(guild_id)
    group_perms = {name: perms for name, perms, _ in squad_groups}

    # Get all active export rows
    # Returns: (whitelist_slug, output_filename, discord_id, discord_name, id_type, id_value)
    rows = await db.get_active_export_rows(guild_id)

    # Build group headers
    def build_group_headers(used_groups: set) -> list[str]:
        lines = []
        for gname in sorted(used_groups):
            perms = group_perms.get(gname, "reserve")
            lines.append(f"Group={gname}:{perms}")
        if lines:
            lines.extend(["", ""])  # Blank lines after group headers
        return lines

    # Build a single admin line
    def build_line(id_type: str, id_value: str, name: str, group_name: str) -> str:
        suffix = " [EOS]" if id_type == "eosid" else ""
        return f"Admin={id_value}:{group_name} // {name}{suffix}"

    outputs = {}

    # Pre-seed group sets from all enabled whitelists so group headers always appear
    # even when a whitelist has zero entries.
    all_enabled_groups: set[str] = set()
    per_wl_default_groups: dict[str, set] = {}
    for wl in whitelists:
        if not wl.get("enabled"):
            continue
        grp = wl.get("squad_group") or "Whitelist"
        all_enabled_groups.add(grp)
        per_wl_default_groups[wl["slug"]] = {grp}

    # Combined mode: all whitelists in one file
    combined_lines = []
    combined_seen = set()
    combined_groups: set[str] = set(all_enabled_groups)  # always include all groups

    # Per-whitelist mode: separate file per whitelist
    per_wl_lines: dict[str, list[str]] = {}
    per_wl_seen: dict[str, set] = {}
    per_wl_groups: dict[str, set] = {slug: set(grps) for slug, grps in per_wl_default_groups.items()}

    for row in rows:
        wl_slug, output_filename, discord_id, discord_name, id_type, id_value = row

        # Find the squad group for this whitelist
        wl = next((w for w in whitelists if w["slug"] == wl_slug), None)
        group_name = wl["squad_group"] if wl else "Whitelist"

        line = build_line(id_type, id_value, discord_name, group_name)
        dedup_key = f"{id_type}:{id_value}" if dedupe else line

        # Combined
        if output_mode in ("combined", "hybrid"):
            if dedup_key not in combined_seen:
                combined_lines.append(line)
                combined_seen.add(dedup_key)
                combined_groups.add(group_name)

        # Separate / Hybrid
        if output_mode in ("separate", "hybrid"):
            if wl_slug not in per_wl_lines:
                per_wl_lines[wl_slug] = []
                per_wl_seen[wl_slug] = set()
            if wl_slug not in per_wl_groups:
                per_wl_groups[wl_slug] = {group_name}

            if dedup_key not in per_wl_seen[wl_slug]:
                per_wl_lines[wl_slug].append(line)
                per_wl_seen[wl_slug].add(dedup_key)
                per_wl_groups[wl_slug].add(group_name)

    # Build combined output
    if output_mode in ("combined", "hybrid"):
        content = build_group_headers(combined_groups) + combined_lines
        combined_content = "\n".join(content)
        outputs[combined_filename] = combined_content
        # Also serve the combined content under each whitelist's own output_filename so
        # per-whitelist URLs always work regardless of output_mode.
        for wl in whitelists:
            if not wl.get("enabled"):
                continue
            wl_fn = wl.get("output_filename")
            if wl_fn and wl_fn != combined_filename:
                outputs[wl_fn] = combined_content

    # Build per-whitelist outputs
    if output_mode in ("separate", "hybrid"):
        for wl in whitelists:
            if not wl["enabled"]:
                continue
            slug = wl["slug"]
            filename = wl["output_filename"]
            lines = per_wl_lines.get(slug, [])
            groups = per_wl_groups.get(slug, set())
            content = build_group_headers(groups) + lines
            outputs[filename] = "\n".join(content)

    return outputs


async def _notify_web_service(guild_id: int) -> None:
    """Fire-and-forget HTTP call to tell the web service to refresh its cache.

    Only called by the bot-worker in two-process deployments where WEB_INTERNAL_URL
    is set.  The web service validates WEB_FILE_SECRET and re-runs sync_outputs.
    """
    import aiohttp
    url = f"{WEB_INTERNAL_URL}/internal/sync/{guild_id}"
    headers = {"Authorization": f"Bearer {WEB_FILE_SECRET}"}
    try:
        async with aiohttp.ClientSession() as session:
            async with session.post(url, headers=headers, timeout=aiohttp.ClientTimeout(total=5)) as resp:
                if resp.status != 200:
                    log.debug("Web service sync notification returned %s for guild %s", resp.status, guild_id)
    except Exception:
        log.debug("Web service sync notification failed for guild %s (web service may be down)", guild_id)


async def sync_outputs(db: "Database", guild_id: int, web_server=None, github=None) -> int:
    """Generate output files and push to web cache + optional GitHub.

    Args:
        db: Database instance
        guild_id: Guild to sync (or None for all guilds)
        web_server: Optional WebServer instance to update cache
        github: Optional GithubPublisher instance

    Returns:
        Number of files changed
    """
    import asyncio

    outputs = await generate_output_files(db, guild_id)

    # Update web server cache (single-process mode)
    if web_server:
        web_server.update_cache(guild_id, outputs)
    elif WEB_INTERNAL_URL:
        # Two-process mode: notify the web service to pull fresh data from DB
        import asyncio
        asyncio.create_task(_notify_web_service(guild_id))

    # Publish to GitHub if configured
    changed = 0
    if github:
        for filename, content in outputs.items():
            try:
                updated = await asyncio.to_thread(github.update_file_if_needed, filename, content)
                if updated:
                    changed += 1
            except Exception:
                log.exception("Failed to sync %s to GitHub", filename)

    return changed

File: api/src/services/output.ts
/**
 * Whitelist output file generator.
 *
 * Builds Squad RemoteAdminList format files from DB data.
 *
 * Each whitelist always gets its own file (separate mode only).
 * Entries are grouped by source with comment headers for readability:
 *   // ─── Discord Roster ───
 *   // ─── Manual Roster ────
 *   // ─── Seeding Rewards ──
 *
 * Format:
 *   Group=Whitelist:reserve
 *   Admin=76561198012345678:Whitelist // PlayerName
 */
import type { PrismaClient } from "@prisma/client"

type OutputMap = Record<string, string>

/**
 * Generate all whitelist output files for a guild.
 * Returns { filename: content } for each file to serve.
 */
export async function syncOutputs(
  prisma: PrismaClient,
  guildId: bigint,
): Promise<OutputMap> {
  const settingsRows = await prisma.botSetting.findMany({
    where: { guildId },
    select: { settingKey: true, settingValue: true },
  })
  const settings = Object.fromEntries(settingsRows.map((r) => [r.settingKey, r.settingValue]))
  const dedupe = toBool(settings["duplicate_output_dedupe"] ?? "true")

  const whitelists = await prisma.whitelist.findMany({
    where: { guildId },
    orderBy: [{ isDefault: "desc" }, { name: "asc" }],
  })

  const groups = await prisma.squadGroup.findMany({ where: { guildId } })
  const groupPerms = Object.fromEntries(groups.map((g) => [g.groupName, g.permissions]))
  const disabledGroups = new Set(groups.filter((g) => !g.enabled).map((g) => g.groupName))

  // Query includes created_via and category for section grouping
  const rows = await prisma.$queryRaw<ExportRow[]>`
    SELECT
      w.slug             AS "wlSlug",
      w.output_filename  AS "outputFilename",
      w.is_manual        AS "isManual",
      u.discord_id       AS "discordId",
      u.discord_name     AS "discordName",
      u.effective_slot_limit AS "slotLimit",
      u.created_via      AS "createdVia",
      u.category_id      AS "categoryId",
      i.id_type          AS "idType",
      i.id_value         AS "idValue",
      c.name             AS "categoryName"
    FROM whitelist_users u
    JOIN whitelists w ON w.id = u.whitelist_id
    JOIN whitelist_identifiers i
      ON u.guild_id = i.guild_id
     AND u.discord_id = i.discord_id
     AND u.whitelist_id = i.whitelist_id
    LEFT JOIN whitelist_categories c ON c.id = u.category_id
    WHERE u.guild_id = ${guildId}
      AND u.status = 'active'
    ORDER BY w.slug, u.created_via, c.name, u.discord_name, i.id_type, i.id_value
  `

  function buildGroupHeaders(usedGroups: Set<string>): string[] {
    const lines: string[] = []
    for (const name of [...usedGroups].sort()) {
      const perms = groupPerms[name] ?? "reserve"
      lines.push(`Group=${name}:${perms}`)
    }
    if (lines.length > 0) lines.push("")
    return lines
  }

  function buildLine(idType: string, idValue: string, name: string, groupName: string): string {
    const suffix = idType === "eosid" ? " [EOS]" : ""
    return `Admin=${idValue}:${groupName} // ${name}${suffix}`
  }

  /**
   * Determine a human-readable section label for a row.
   * Used to group entries with comment headers.
   */
  function getSectionLabel(row: ExportRow): string {
    if (row.createdVia === "seeding_reward") return "Seeding Rewards"
    if (row.isManual) {
      return row.categoryName ? `Manual Roster - ${row.categoryName}` : "Manual Roster"
    }
    return "Discord Roster"
  }

  // ── Build per-whitelist output ──────────────────────────────────────────

  const enabledWhitelists = whitelists.filter(
    (w) => w.enabled && !disabledGroups.has(w.squadGroup),
  )
  const wlBySlug = Object.fromEntries(whitelists.map((w) => [w.slug, w]))

  // Structured output: section -> lines
  const perWlSections: Record<string, Map<string, string[]>> = {}
  const perWlSeen:     Record<string, Set<string>> = {}
  const perWlGroups:   Record<string, Set<string>> = {}
  const userIdCounts = new Map<string, number>()

  for (const wl of enabledWhitelists) {
    perWlGroups[wl.slug] = new Set([wl.squadGroup])
  }

  for (const row of rows) {
    const wl = wlBySlug[row.wlSlug]
    if (!wl || !wl.enabled) continue

    const userKey = `${row.wlSlug}:${String(row.discordId)}`
    const exported = userIdCounts.get(userKey) ?? 0
    if (row.slotLimit > 0 && exported >= row.slotLimit) continue

    // Seeding rewards always use Reserve group regardless of whitelist's group
    const groupName = row.createdVia === "seeding_reward" ? "Reserve" : (wl.squadGroup || "Whitelist")
    const line      = buildLine(row.idType, row.idValue, row.discordName, groupName)
    const dedupKey  = dedupe ? `${row.idType}:${row.idValue}` : line
    const slug      = row.wlSlug
    const section   = getSectionLabel(row)

    perWlSections[slug] ??= new Map()
    perWlSeen[slug]     ??= new Set()
    perWlGroups[slug]   ??= new Set([groupName])

    if (!perWlSeen[slug].has(dedupKey)) {
      if (!perWlSections[slug].has(section)) {
        perWlSections[slug].set(section, [])
      }
      perWlSections[slug].get(section)!.push(line)
      perWlSeen[slug].add(dedupKey)
      perWlGroups[slug].add(groupName)
      userIdCounts.set(userKey, exported + 1)
    }
  }

  // ── Assemble outputs with section headers ─────────────────────────────

  const outputs: OutputMap = {}

  for (const wl of enabledWhitelists) {
    const filename = wl.outputFilename || `${wl.slug}.txt`
    const wlGroups = perWlGroups[wl.slug] ?? new Set([wl.squadGroup])
    const sections = perWlSections[wl.slug]

    const allLines: string[] = [...buildGroupHeaders(wlGroups)]

    if (sections && sections.size > 0) {
      for (const [sectionName, lines] of sections) {
        // Only add section headers if there are multiple sections
        if (sections.size > 1) {
          allLines.push(`// ─── ${sectionName} ${"─".repeat(Math.max(0, 40 - sectionName.length))}`)
        }
        allLines.push(...lines)
        if (sections.size > 1) allLines.push("")
      }
    }

    outputs[filename] = allLines.join("\n")
  }

  return outputs
}

// ─── Helpers ──────────────────────────────────────────────────────────────────

interface ExportRow {
  wlSlug:         string
  outputFilename: string
  isManual:       boolean
  discordId:      bigint
  discordName:    string
  slotLimit:      number
  createdVia:     string | null
  categoryId:     number | null
  idType:         string
  idValue:        string
  categoryName:   string | null
}

function toBool(val: string): boolean {
  return ["1", "true", "yes", "on"].includes(val.toLowerCase())
}
